File: backend/routers/rip/rip.py
"""RIP Protocol Router.

API endpoints for managing VyOS RIP (Routing Information Protocol) configuration.
Supports version-aware configuration for VyOS 1.4 and 1.5.
"""

from fastapi import APIRouter, HTTPException, Request
from starlette.concurrency import run_in_threadpool
from pydantic import BaseModel, Field
from typing import List, Dict, Optional, Any
from session_vyos_service import get_session_vyos_service
from vyos_builders import RipBatchBuilder
from fastapi_permissions import require_read_permission, require_write_permission
from rbac_permissions import FeatureGroup
import inspect
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/vyos/rip", tags=["rip"])


# ============================================================================
# Pydantic Models
# ============================================================================


class RipDistributeListGlobal(BaseModel):
    access_list_in: Optional[str] = None
    access_list_out: Optional[str] = None
    prefix_list_in: Optional[str] = None
    prefix_list_out: Optional[str] = None


class RipDistributeListInterface(BaseModel):
    interface: str
    access_list_in: Optional[str] = None
    access_list_out: Optional[str] = None
    prefix_list_in: Optional[str] = None
    prefix_list_out: Optional[str] = None


class RipDistributeList(BaseModel):
    global_filters: RipDistributeListGlobal = RipDistributeListGlobal()
    interface_filters: List[RipDistributeListInterface] = []


class RipMd5Key(BaseModel):
    key_id: str
    password: str


class RipInterface(BaseModel):
    name: str
    authentication_type: Optional[str] = None
    md5_keys: List[RipMd5Key] = []
    plaintext_password: Optional[str] = None
    receive_version: Optional[str] = None
    send_version: Optional[str] = None
    split_horizon: Optional[str] = None


class RipNetworkDistance(BaseModel):
    prefix: str
    distance: Optional[int] = None
    access_list: Optional[str] = None


class RipRedistribute(BaseModel):
    protocol: str
    metric: Optional[int] = None
    route_map: Optional[str] = None


class RipTimers(BaseModel):
    update: Optional[int] = None
    timeout: Optional[int] = None
    garbage_collection: Optional[int] = None


class RipConfig(BaseModel):
    default_distance: Optional[int] = None
    default_information_originate: bool = False
    default_metric: Optional[int] = None
    route_map: Optional[str] = None
    version: Optional[str] = None
    networks: List[str] = []
    neighbors: List[str] = []
    routes: List[str] = []
    passive_interfaces: List[str] = []
    distribute_list: RipDistributeList = RipDistributeList()
    interfaces: List[RipInterface] = []
    network_distances: List[RipNetworkDistance] = []
    redistribute: List[RipRedistribute] = []
    timers: RipTimers = RipTimers()


class RipBatchOperation(BaseModel):
    op: str = Field(..., description="Builder method name")
    value: Optional[str] = Field(None, description="Comma-separated arguments")


class RipBatchRequest(BaseModel):
    operations: List[RipBatchOperation]


class VyOSResponse(BaseModel):
    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


# ============================================================================
# Endpoint 1: Capabilities
# ============================================================================


@router.get("/capabilities")
async def get_rip_capabilities(request: Request):
    """Return RIP feature capabilities based on the device VyOS version."""
    await require_read_permission(request, FeatureGroup.RIP)

    try:
        service = get_session_vyos_service(request)
        builder = RipBatchBuilder(version=service.get_version())
        capabilities = builder.get_capabilities()

        if hasattr(request.state, "instance") and request.state.instance:
            capabilities["instance_name"] = request.state.instance.get("name")
            capabilities["instance_id"] = request.state.instance.get("id")

        return capabilities
    except Exception:
        logger.exception("Unhandled error in get_rip_capabilities")
        raise HTTPException(status_code=500, detail="Internal server error")


# ============================================================================
# Endpoint 2: Config
# ============================================================================


@router.get("/config", response_model=RipConfig)
async def get_rip_config(http_request: Request, refresh: bool = False):
    """Return the full RIP configuration in a normalized format."""
    await require_read_permission(http_request, FeatureGroup.RIP)

    try:
        service = get_session_vyos_service(http_request)
        full_config = await run_in_threadpool(service.get_full_config, refresh=refresh)

        rip_raw = full_config.get("protocols", {}).get("rip", {})

        if not rip_raw:
            return RipConfig()

        return RipConfig(
            default_distance=_safe_int(rip_raw.get("default-distance")),
            default_information_originate="originate" in (rip_raw.get("default-information") or {}),
            default_metric=_safe_int(rip_raw.get("default-metric")),
            route_map=rip_raw.get("route-map"),
            version=rip_raw.get("version"),
            networks=_to_list(rip_raw.get("network")),
            neighbors=_to_list(rip_raw.get("neighbor")),
            routes=_to_list(rip_raw.get("route")),
            passive_interfaces=_to_list(rip_raw.get("passive-interface")),
            distribute_list=_parse_distribute_list(rip_raw.get("distribute-list", {})),
            interfaces=_parse_interfaces(rip_raw.get("interface", {})),
            network_distances=_parse_network_distances(rip_raw.get("network-distance", {})),
            redistribute=_parse_redistribute(rip_raw.get("redistribute", {})),
            timers=_parse_timers(rip_raw.get("timers", {})),
        )
    except Exception:
        logger.exception("Unhandled error in get_rip_config")
        raise HTTPException(status_code=500, detail="Internal server error")


# ============================================================================
# Config Parsers
# ============================================================================


def _safe_int(value) -> Optional[int]:
    if value is None:
        return None
    try:
        return int(value)
    except (ValueError, TypeError):
        return None


def _to_list(value) -> List[str]:
    if value is None:
        return []
    if isinstance(value, list):
        return value
    if isinstance(value, str):
        return [value]
    if isinstance(value, dict):
        return list(value.keys())
    return []


def _parse_distribute_list(raw: dict) -> RipDistributeList:
    if not raw:
        return RipDistributeList()

    acl_raw = raw.get("access-list", {}) or {}
    pl_raw = raw.get("prefix-list", {}) or {}

    global_filters = RipDistributeListGlobal(
        access_list_in=acl_raw.get("in") if isinstance(acl_raw, dict) else None,
        access_list_out=acl_raw.get("out") if isinstance(acl_raw, dict) else None,
        prefix_list_in=pl_raw.get("in") if isinstance(pl_raw, dict) else None,
        prefix_list_out=pl_raw.get("out") if isinstance(pl_raw, dict) else None,
    )

    iface_filters: List[RipDistributeListInterface] = []
    for iface, iface_raw in (raw.get("interface", {}) or {}).items():
        if iface_raw is None:
            iface_raw = {}
        iface_acl = iface_raw.get("access-list", {}) or {}
        iface_pl = iface_raw.get("prefix-list", {}) or {}
        iface_filters.append(RipDistributeListInterface(
            interface=iface,
            access_list_in=iface_acl.get("in") if isinstance(iface_acl, dict) else None,
            access_list_out=iface_acl.get("out") if isinstance(iface_acl, dict) else None,
            prefix_list_in=iface_pl.get("in") if isinstance(iface_pl, dict) else None,
            prefix_list_out=iface_pl.get("out") if isinstance(iface_pl, dict) else None,
        ))

    return RipDistributeList(global_filters=global_filters, interface_filters=iface_filters)


def _parse_interfaces(raw: dict) -> List[RipInterface]:
    if not raw:
        return []

    interfaces = []
    for iface_name, cfg in raw.items():
        if cfg is None:
            cfg = {}

        auth_raw = cfg.get("authentication", {}) or {}
        md5_raw = auth_raw.get("md5", {}) or {}
        plaintext = auth_raw.get("plaintext-password")

        md5_keys = []
        for key_id, key_cfg in md5_raw.items():
            if key_cfg is None:
                key_cfg = {}
            md5_keys.append(RipMd5Key(
                key_id=str(key_id),
                password=key_cfg.get("password", "") if isinstance(key_cfg, dict) else "",
            ))

        auth_type = None
        if md5_keys:
            auth_type = "md5"
        elif plaintext:
            auth_type = "plaintext"

        split_horizon_raw = cfg.get("split-horizon", {}) or {}
        split_horizon = None
        if "poison-reverse" in split_horizon_raw:
            split_horizon = "poison-reverse"
        elif "disable" in split_horizon_raw:
            split_horizon = "disable"

        interfaces.append(RipInterface(
            name=iface_name,
            authentication_type=auth_type,
            md5_keys=md5_keys,
            plaintext_password=plaintext,
            receive_version=(cfg.get("receive", {}) or {}).get("version"),
            send_version=(cfg.get("send", {}) or {}).get("version"),
            split_horizon=split_horizon,
        ))

    return interfaces


def _parse_network_distances(raw: dict) -> List[RipNetworkDistance]:
    if not raw:
        return []

    entries = []
    for prefix, cfg in raw.items():
        if cfg is None:
            cfg = {}
        entries.append(RipNetworkDistance(
            prefix=prefix,
            distance=_safe_int(cfg.get("distance")),
            access_list=cfg.get("access-list"),
        ))
    return entries


def _parse_redistribute(raw: dict) -> List[RipRedistribute]:
    if not raw:
        return []

    entries = []
    for protocol, cfg in raw.items():
        if cfg is None:
            cfg = {}
        entries.append(RipRedistribute(
            protocol=protocol,
            metric=_safe_int(cfg.get("metric")),
            route_map=cfg.get("route-map"),
        ))
    return entries


def _parse_timers(raw: dict) -> RipTimers:
    if not raw:
        return RipTimers()
    return RipTimers(
        update=_safe_int(raw.get("update")),
        timeout=_safe_int(raw.get("timeout")),
        garbage_collection=_safe_int(raw.get("garbage-collection")),
    )


# ============================================================================
# Endpoint 3: Batch Operations
# ============================================================================


@router.post("/batch", response_model=VyOSResponse)
async def rip_batch_configure(http_request: Request, body: RipBatchRequest):
    """Execute a batch of RIP configuration operations atomically."""
    await require_write_permission(http_request, FeatureGroup.RIP)

    try:
        service = get_session_vyos_service(http_request)
        builder = RipBatchBuilder(version=service.get_version())

        for operation in body.operations:
            method = getattr(builder, operation.op)
            sig = inspect.signature(method)
            params = [p for p in sig.parameters.keys() if p != "self"]

            if len(params) == 0:
                method()
            elif len(params) == 1:
                if operation.value is not None:
                    method(operation.value)
            elif len(params) == 2 and operation.value is not None:
                values = operation.value.split(",", 1)
                if len(values) == 2:
                    method(values[0], values[1])
                else:
                    method(operation.value, "")
            elif len(params) == 3 and operation.value is not None:
                values = operation.value.split(",", 2)
                if len(values) == 3:
                    method(values[0], values[1], values[2])
                elif len(values) == 2:
                    method(values[0], values[1], "")

        response = service.execute_batch(builder)

        return VyOSResponse(
            success=response.status == 200,
            data={"message": "RIP configuration updated"},
            error=response.error if response.error else None,
        )
    except AttributeError as e:
        raise HTTPException(status_code=400, detail=f"Unknown operation: {str(e)}")
    except Exception:
        logger.exception("Unhandled error in rip_batch_configure")
        raise HTTPException(status_code=500, detail="Internal server error")

File: frontend/src/lib/api/rip.ts
import { apiClient } from "./client";

// ============================================================================
// TypeScript Interfaces
// ============================================================================

export interface RipDistributeListGlobal {
  access_list_in?: string | null;
  access_list_out?: string | null;
  prefix_list_in?: string | null;
  prefix_list_out?: string | null;
}

export interface RipDistributeListInterface {
  interface: string;
  access_list_in?: string | null;
  access_list_out?: string | null;
  prefix_list_in?: string | null;
  prefix_list_out?: string | null;
}

export interface RipDistributeList {
  global_filters: RipDistributeListGlobal;
  interface_filters: RipDistributeListInterface[];
}

export interface RipMd5Key {
  key_id: string;
  password: string;
}

export interface RipInterface {
  name: string;
  authentication_type?: string | null;
  md5_keys: RipMd5Key[];
  plaintext_password?: string | null;
  receive_version?: string | null;
  send_version?: string | null;
  split_horizon?: string | null;
}

export interface RipNetworkDistance {
  prefix: string;
  distance?: number | null;
  access_list?: string | null;
}

export interface RipRedistribute {
  protocol: string;
  metric?: number | null;
  route_map?: string | null;
}

export interface RipTimers {
  update?: number | null;
  timeout?: number | null;
  garbage_collection?: number | null;
}

export interface RipConfig {
  default_distance?: number | null;
  default_information_originate: boolean;
  default_metric?: number | null;
  route_map?: string | null;
  version?: string | null;
  networks: string[];
  neighbors: string[];
  routes: string[];
  passive_interfaces: string[];
  distribute_list: RipDistributeList;
  interfaces: RipInterface[];
  network_distances: RipNetworkDistance[];
  redistribute: RipRedistribute[];
  timers: RipTimers;
}

export interface RipCapabilities {
  version: string;
  version_info: {
    is_1_4: boolean;
    is_1_5: boolean;
  };
  features: {
    global_settings: { supported: boolean; description: string };
    networks: { supported: boolean; description: string };
    neighbors: { supported: boolean; description: string };
    static_routes: { supported: boolean; description: string };
    passive_interfaces: { supported: boolean; description: string };
    distribute_lists: { supported: boolean; description: string };
    interface_settings: { supported: boolean; description: string };
    network_distance: { supported: boolean; description: string };
    redistribute: { supported: boolean; description: string; protocols: string[] };
    timers: { supported: boolean; description: string };
  };
  instance_name?: string;
  instance_id?: string;
}

export interface RipBatchOperation {
  op: string;
  value?: string;
}

export interface RipBatchRequest {
  operations: RipBatchOperation[];
}

export interface VyOSResponse {
  success: boolean;
  data?: Record<string, unknown> | null;
  error?: string | null;
}

// ============================================================================
// API Service
// ============================================================================

class RipService {
  async getCapabilities(): Promise<RipCapabilities> {
    return apiClient.get<RipCapabilities>("/vyos/rip/capabilities");
  }

  async getConfig(refresh = false): Promise<RipConfig> {
    return apiClient.get<RipConfig>("/vyos/rip/config", {
      refresh: refresh.toString(),
    });
  }

  async refreshConfig(): Promise<VyOSResponse> {
    return apiClient.post<VyOSResponse>("/vyos/config/refresh");
  }

  async batchConfigure(request: RipBatchRequest): Promise<VyOSResponse> {
    const result = await apiClient.post<VyOSResponse>("/vyos/rip/batch", request);
    if (!result.success) throw new Error(result.error || "Operation failed");
    await this.refreshConfig();
    return result;
  }

  // ==========================================================================
  // Global Settings
  // ==========================================================================

  async updateGlobalSettings(
    original: RipConfig,
    updated: Partial<RipConfig>
  ): Promise<VyOSResponse> {
    const ops: RipBatchOperation[] = [];

    if (updated.version !== original.version) {
      ops.push(
        updated.version
          ? { op: "set_version", value: updated.version }
          : { op: "delete_version" }
      );
    }

    if (updated.default_distance !== original.default_distance) {
      ops.push(
        updated.default_distance != null
          ? { op: "set_default_distance", value: String(updated.default_distance) }
          : { op: "delete_default_distance" }
      );
    }

    if (updated.default_metric !== original.default_metric) {
      ops.push(
        updated.default_metric != null
          ? { op: "set_default_metric", value: String(updated.default_metric) }
          : { op: "delete_default_metric" }
      );
    }

    if (updated.route_map !== original.route_map) {
      ops.push(
        updated.route_map
          ? { op: "set_route_map", value: updated.route_map }
          : { op: "delete_route_map" }
      );
    }

    if (
      updated.default_information_originate !== undefined &&
      updated.default_information_originate !== original.default_information_originate
    ) {
      ops.push(
        updated.default_information_originate
          ? { op: "set_default_information_originate" }
          : { op: "delete_default_information_originate" }
      );
    }

    if (ops.length === 0) return { success: true, data: null };
    return this.batchConfigure({ operations: ops });
  }

  async updateTimers(
    original: RipTimers,
    updated: RipTimers
  ): Promise<VyOSResponse> {
    const ops: RipBatchOperation[] = [];

    if (updated.update !== original.update) {
      ops.push(
        updated.update != null
          ? { op: "set_timers_update", value: String(updated.update) }
          : { op: "delete_timers_update" }
      );
    }

    if (updated.timeout !== original.timeout) {
      ops.push(
        updated.timeout != null
          ? { op: "set_timers_timeout", value: String(updated.timeout) }
          : { op: "delete_timers_timeout" }
      );
    }

    if (updated.garbage_collection !== original.garbage_collection) {
      ops.push(
        updated.garbage_collection != null
          ? { op: "set_timers_garbage_collection", value: String(updated.garbage_collection) }
          : { op: "delete_timers_garbage_collection" }
      );
    }

    if (ops.length === 0) return { success: true, data: null };
    return this.batchConfigure({ operations: ops });
  }

  // ==========================================================================
  // Networks / Neighbors / Routes / Passive Interfaces
  // ==========================================================================

  async addNetwork(network: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "set_network", value: network }] });
  }

  async removeNetwork(network: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "delete_network", value: network }] });
  }

  async addNeighbor(address: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "set_neighbor", value: address }] });
  }

  async removeNeighbor(address: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "delete_neighbor", value: address }] });
  }

  async addRoute(prefix: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "set_route", value: prefix }] });
  }

  async removeRoute(prefix: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "delete_route", value: prefix }] });
  }

  async addPassiveInterface(iface: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "set_passive_interface", value: iface }] });
  }

  async removePassiveInterface(iface: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "delete_passive_interface", value: iface }] });
  }

  // ==========================================================================
  // Interface Settings
  // ==========================================================================

  async createInterface(config: RipInterface): Promise<VyOSResponse> {
    const ops: RipBatchOperation[] = [{ op: "set_interface", value: config.name }];
    this._buildInterfaceOps(ops, config);
    return this.batchConfigure({ operations: ops });
  }

  async updateInterface(
    original: RipInterface,
    updated: RipInterface
  ): Promise<VyOSResponse> {
    const ops: RipBatchOperation[] = [];
    const name = original.name;

    // Authentication — delete old, set new
    const origAuthType = original.authentication_type;
    const newAuthType = updated.authentication_type;

    if (origAuthType !== newAuthType || JSON.stringify(original.md5_keys) !== JSON.stringify(updated.md5_keys) || original.plaintext_password !== updated.plaintext_password) {
      if (origAuthType) {
        ops.push({ op: "delete_interface_authentication", value: name });
      }
      if (newAuthType === "md5") {
        for (const key of updated.md5_keys) {
          ops.push({ op: "set_interface_authentication_md5_key", value: `${name},${key.key_id},${key.password}` });
        }
      } else if (newAuthType === "plaintext" && updated.plaintext_password) {
        ops.push({ op: "set_interface_authentication_plaintext", value: `${name},${updated.plaintext_password}` });
      }
    }

    if (updated.send_version !== original.send_version) {
      ops.push(
        updated.send_version
          ? { op: "set_interface_send_version", value: `${name},${updated.send_version}` }
          : { op: "delete_interface_send_version", value: name }
      );
    }

    if (updated.receive_version !== original.receive_version) {
      ops.push(
        updated.receive_version
          ? { op: "set_interface_receive_version", value: `${name},${updated.receive_version}` }
          : { op: "delete_interface_receive_version", value: name }
      );
    }

    if (updated.split_horizon !== original.split_horizon) {
      if (original.split_horizon) {
        ops.push({ op: "delete_interface_split_horizon", value: name });
      }
      if (updated.split_horizon === "disable") {
        ops.push({ op: "set_interface_split_horizon_disable", value: name });
      } else if (updated.split_horizon === "poison-reverse") {
        ops.push({ op: "set_interface_split_horizon_poison_reverse", value: name });
      }
    }

    if (ops.length === 0) return { success: true, data: null };
    return this.batchConfigure({ operations: ops });
  }

  async deleteInterface(name: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "delete_interface", value: name }] });
  }

  private _buildInterfaceOps(ops: RipBatchOperation[], config: RipInterface): void {
    const name = config.name;
    if (config.authentication_type === "md5") {
      for (const key of config.md5_keys) {
        ops.push({ op: "set_interface_authentication_md5_key", value: `${name},${key.key_id},${key.password}` });
      }
    } else if (config.authentication_type === "plaintext" && config.plaintext_password) {
      ops.push({ op: "set_interface_authentication_plaintext", value: `${name},${config.plaintext_password}` });
    }
    if (config.send_version) {
      ops.push({ op: "set_interface_send_version", value: `${name},${config.send_version}` });
    }
    if (config.receive_version) {
      ops.push({ op: "set_interface_receive_version", value: `${name},${config.receive_version}` });
    }
    if (config.split_horizon === "disable") {
      ops.push({ op: "set_interface_split_horizon_disable", value: name });
    } else if (config.split_horizon === "poison-reverse") {
      ops.push({ op: "set_interface_split_horizon_poison_reverse", value: name });
    }
  }

  // ==========================================================================
  // Redistribute
  // ==========================================================================

  async createRedistribute(entry: RipRedistribute): Promise<VyOSResponse> {
    const ops: RipBatchOperation[] = [{ op: "set_redistribute", value: entry.protocol }];
    if (entry.metric != null) {
      ops.push({ op: "set_redistribute_metric", value: `${entry.protocol},${entry.metric}` });
    }
    if (entry.route_map) {
      ops.push({ op: "set_redistribute_route_map", value: `${entry.protocol},${entry.route_map}` });
    }
    return this.batchConfigure({ operations: ops });
  }

  async updateRedistribute(
    original: RipRedistribute,
    updated: RipRedistribute
  ): Promise<VyOSResponse> {
    const ops: RipBatchOperation[] = [];
    const proto = original.protocol;

    if (updated.metric !== original.metric) {
      ops.push(
        updated.metric != null
          ? { op: "set_redistribute_metric", value: `${proto},${updated.metric}` }
          : { op: "set_redistribute", value: proto }
      );
    }

    if (updated.route_map !== original.route_map) {
      if (updated.route_map) {
        ops.push({ op: "set_redistribute_route_map", value: `${proto},${updated.route_map}` });
      } else {
        ops.push({ op: "delete_redistribute", value: proto });
        ops.push({ op: "set_redistribute", value: proto });
        if (updated.metric != null) {
          ops.push({ op: "set_redistribute_metric", value: `${proto},${updated.metric}` });
        }
      }
    }

    if (ops.length === 0) return { success: true, data: null };
    return this.batchConfigure({ operations: ops });
  }

  async deleteRedistribute(protocol: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "delete_redistribute", value: protocol }] });
  }

  // ==========================================================================
  // Network Distance
  // ==========================================================================

  async createNetworkDistance(entry: RipNetworkDistance): Promise<VyOSResponse> {
    const ops: RipBatchOperation[] = [{ op: "set_network_distance", value: entry.prefix }];
    if (entry.distance != null) {
      ops.push({ op: "set_network_distance_value", value: `${entry.prefix},${entry.distance}` });
    }
    if (entry.access_list) {
      ops.push({ op: "set_network_distance_access_list", value: `${entry.prefix},${entry.access_list}` });
    }
    return this.batchConfigure({ operations: ops });
  }

  async updateNetworkDistance(
    original: RipNetworkDistance,
    updated: RipNetworkDistance
  ): Promise<VyOSResponse> {
    const ops: RipBatchOperation[] = [];
    const prefix = original.prefix;

    if (updated.distance !== original.distance) {
      if (updated.distance != null) {
        ops.push({ op: "set_network_distance_value", value: `${prefix},${updated.distance}` });
      }
    }

    if (updated.access_list !== original.access_list) {
      if (updated.access_list) {
        ops.push({ op: "set_network_distance_access_list", value: `${prefix},${updated.access_list}` });
      } else {
        ops.push({ op: "delete_network_distance", value: prefix });
        ops.push({ op: "set_network_distance", value: prefix });
        if (updated.distance != null) {
          ops.push({ op: "set_network_distance_value", value: `${prefix},${updated.distance}` });
        }
      }
    }

    if (ops.length === 0) return { success: true, data: null };
    return this.batchConfigure({ operations: ops });
  }

  async deleteNetworkDistance(prefix: string): Promise<VyOSResponse> {
    return this.batchConfigure({ operations: [{ op: "delete_network_distance", value: prefix }] });
  }

  // ==========================================================================
  // Distribute List - Global
  // ==========================================================================

  async updateDistributeListGlobal(
    original: RipDistributeListGlobal,
    updated: RipDistributeListGlobal
  ): Promise<VyOSResponse> {
    const ops: RipBatchOperation[] = [];

    if (updated.access_list_in !== original.access_list_in) {
      ops.push(
        updated.access_list_in
          ? { op: "set_distribute_list_access_list_in", value: updated.access_list_in }
          : { op: "delete_distribute_list_access_list_in" }
      );
    }

    if (updated.access_list_out !== original.access_list_out) {
      ops.push(
        updated.access_list_out
          ? { op: "set_distribute_list_access_list_out", value: updated.access_list_out }
          : { op: "delete_distribute_list_access_list_out" }
      );
    }

    if (updated.prefix_list_in !== original.prefix_list_in) {
      ops.push(
        updated.prefix_list_in
          ? { op: "set_distribute_list_prefix_list_in", value: updated.prefix_list_in }
          : { op: "delete_distribute_list_prefix_list_in" }
      );
    }

    if (updated.prefix_list_out !== original.prefix_list_out) {
      ops.push(
        updated.prefix_list_out
          ? { op: "set_distribute_list_prefix_list_out", value: updated.prefix_list_out }
          : { op: "delete_distribute_list_prefix_list_out" }
      );
    }

    if (ops.length === 0) return { success: true, data: null };
    return this.batchConfigure({ operations: ops });
  }

  // ==========================================================================
  // Distribute List - Per Interface
  // ==========================================================================

  async createDistributeListInterface(
    entry: RipDistributeListInterface
  ): Promise<VyOSResponse> {
    const ops: RipBatchOperation[] = [];
    const iface = entry.interface;
    if (entry.access_list_in) {
      ops.push({ op: "set_distribute_list_interface_access_list_in", value: `${iface},${entry.access_list_in}` });
    }
    if (entry.access_list_out) {
      ops.push({ op: "set_distribute_list_interface_access_list_out", value: `${iface},${entry.access_list_out}` });
    }
    if (entry.prefix_list_in) {
      ops.push({ op: "set_distribute_list_interface_prefix_list_in", value: `${iface},${entry.prefix_list_in}` });
    }
    if (entry.prefix_list_out) {
      ops.push({ op: "set_distribute_list_interface_prefix_list_out", value: `${iface},${entry.prefix_list_out}` });
    }
    if (ops.length === 0) return { success: true, data: null };
    return this.batchConfigure({ operations: ops });
  }

  async updateDistributeListInterface(
    original: RipDistributeListInterface,
    updated: RipDistributeListInterface
  ): Promise<VyOSResponse> {
    const ops: RipBatchOperation[] = [];
    const iface = original.interface;

    if (updated.access_list_in !== original.access_list_in) {
      ops.push(
        updated.access_list_in
          ? { op: "set_distribute_list_interface_access_list_in", value: `${iface},${updated.access_list_in}` }
          : { op: "delete_distribute_list_interface_access_list_in", value: iface }
      );
    }

    if (updated.access_list_out !== original.access_list_out) {
      ops.push(
        updated.access_list_out
          ? { op: "set_distribute_list_interface_access_list_out", value: `${iface},${updated.access_list_out}` }
          : { op: "delete_distribute_list_interface_access_list_out", value: iface }
      );
    }

    if (updated.prefix_list_in !== original.prefix_list_in) {
      ops.push(
        updated.prefix_list_in
          ? { op: "set_distribute_list_interface_prefix_list_in", value: `${iface},${updated.prefix_list_in}` }
          : { op: "delete_distribute_list_interface_prefix_list_in", value: iface }
      );
    }

    if (updated.prefix_list_out !== original.prefix_list_out) {
      ops.push(
        updated.prefix_list_out
          ? { op: "set_distribute_list_interface_prefix_list_out", value: `${iface},${updated.prefix_list_out}` }
          : { op: "delete_distribute_list_interface_prefix_list_out", value: iface }
      );
    }

    if (ops.length === 0) return { success: true, data: null };
    return this.batchConfigure({ operations: ops });
  }

  async deleteDistributeListInterface(iface: string): Promise<VyOSResponse> {
    return this.batchConfigure({
      operations: [{ op: "delete_distribute_list_interface", value: iface }],
    });
  }
}

export const ripService = new RipService();
